aea/test_tools/test_protocol.py

```diff
@@ -0,0 +1,169 @@
+
+        actual_envelope = Envelope.decode(envelope_bytes)
+        expected_envelope = envelope
+
+        assert expected_envelope.to == actual_envelope.to
+        assert expected_envelope.sender == actual_envelope.sender
+        assert (
+            expected_envelope.protocol_specification_id
+            == actual_envelope.protocol_specification_id
+        )
+        assert expected_envelope.message != actual_envelope.message
+
+        actual_msg = self.MESSAGE_CLASS.serializer.decode(actual_envelope.message_bytes)
+        actual_msg.to = actual_envelope.to
+        actual_msg.sender = actual_envelope.sender
+        expected_msg = msg
+        assert expected_msg == actual_msg
+
+
+class BaseProtocolMessageConstructionTestCase(BaseMessageTestCase):
+    """Base class to test message construction for the protocol."""
+
+    def test_run(self) -> None:
+        """Run the test."""
+        msg = self.build_message()
+        self.perform_message_test(msg)
+
+    @abstractmethod
+    def build_message(self) -> Message:
+        """Build the message to be used for testing."""
+
+
+class BaseProtocolMessagesTestCase(BaseMessageTestCase):
+    """Base class to test messages for the protocol."""
+
+    def test_messages_ok(self) -> None:
+        """Run messages are ok for encode and decode."""
+        for msg in self.build_messages():
+            self.perform_message_test(msg)
+
+    def test_messages_inconsistent(self) -> None:
+        """Run messages are inconsistent."""
+        for msg in self.build_inconsistent():
+            assert not msg._is_consistent()  # pylint: disable=protected-access # nosec
+
+    def test_messages_fail_to_encode_decode(self) -> None:
+        """Run messages are failing to encode and decode."""
+        msg = self.build_messages()[0]
+        with patch.object(msg.__class__.Performative, "__eq__", return_value=False):
+            with pytest.raises(
+                ValueError, match=f"Performative not valid: {msg.performative}"
+            ):
+                msg.serializer.encode(msg)
+
+        encoded_msg = msg.serializer.encode(msg)
+        with patch.object(msg.__class__.Performative, "__eq__", return_value=False):
+            with pytest.raises(
+                ValueError, match=f"Performative not valid: {msg.performative}"
+            ):
+                msg.serializer.decode(encoded_msg)
+
+    @abstractmethod
+    def build_messages(self) -> List[Message]:
+        """Build the messages to be used for testing."""
+
+    @abstractmethod
+    def build_inconsistent(self) -> List[Message]:
+        """Build inconsistent messages to be used for testing."""
+
+
+class BaseProtocolDialoguesTestCase(TestCase):
+    """Base class to test message construction for the protocol."""
+
+    __test__ = False
+    MESSAGE_CLASS: Type[Message] = Message
+    DIALOGUE_CLASS: Type[Dialogue] = Dialogue
+    DIALOGUES_CLASS: Type[Dialogues] = Dialogues
+    ROLE_FOR_THE_FIRST_MESSAGE: Dialogue.Role
```

Comment:
why assign defaults for `MESSAGE_CLASS`, `DIALOGUE_CLASS` and `DIALOGUES_CLASS`? Seems to me it's better not to do this, it will need to be overwritten and without the defaults an error here will be caught earlier. You could add a `setup_test` method on the base class that checks that these attributes are assigned at class level.